src/main/java/frc/robot/PoseEstimation.java

```diff
@@ -0,0 +1,48 @@
+package frc.robot;
+
+import edu.wpi.first.math.VecBuilder;
+import edu.wpi.first.math.estimator.SwerveDrivePoseEstimator;
+import edu.wpi.first.math.geometry.Pose2d;
+import frc.robot.swerve.SwerveDrive;
+import frc.robot.vision.Vision;
+
+public class PoseEstimation {
+    private static PoseEstimation INSTANCE = null;
+    private SwerveDrivePoseEstimator estimator;
+    private Vision vision = Vision.getInstance();
+    private SwerveDrive swerveDrive = SwerveDrive.getInstance();
+
+    public PoseEstimation() {
+        estimator =
+                new SwerveDrivePoseEstimator(
+                        swerveDrive.getKinematics(),
+                        swerveDrive.getYaw(),
+                        swerveDrive.getModulePositions(),
+                        swerveDrive.getBotPose());
+    }
+
+    public static PoseEstimation getInstance() {
+        if (INSTANCE == null) {
+            INSTANCE = new PoseEstimation();
+        }
+        return INSTANCE;
+    }
+
+    public void addVisionMeasurement() {
+        var results = vision.getResults();
```

Comment:
This is something I noticed only now, but `Vision.getResults()` returns the PhotonLib class, while it should return a generic class. Something to fix in the off-season.

Reply:
Will do